packages/react/fast-components-react-base/src/stack-panel/stack-panel.tsx

```diff
@@ -0,0 +1,570 @@
+
+        for (let i: number = 0; i < itemCount; i++) {
+            const itemSpan: number = Array.isArray(this.props.itemSpan)
+                ? this.props.itemSpan[i]
+                : this.props.itemSpan;
+            const currentItemBottom: number = currentItemStart + itemSpan;
+            itemPositions.push({
+                span: itemSpan,
+                start: currentItemStart,
+                end: currentItemBottom,
+            });
+
+            currentItemStart = currentItemBottom;
+        }
+
+        this.itemPositions = itemPositions;
+        this.itemContainerSpan =
+            this.itemPositions.length > 0 ? this.itemPositions[itemCount - 1].end : 0;
+
+        this.maxScroll = this.getMaxScrollDistance();
+    };
+
+    /**
+     * Gets the maximum distance that can be scrolled
+     */
+    private getMaxScrollDistance(): number {
+        if (this.itemPositions.length === 0) {
+            return 0;
+        }
+
+        const contentSpan: number = this.itemPositions[this.itemPositions.length - 1].end;
+        return contentSpan - this.viewportSpan;
+    }
+
+    /**
+     *  Get the size of the viewport along the panel's axis
+     */
+    private getViewportSpan = (): number => {
+        if (isNil(this.rootElement.current)) {
+            return 0;
+        }
+        if (this.props.orientation === Orientation.horizontal) {
+            return this.rootElement.current.clientWidth;
+        } else {
+            return this.rootElement.current.clientHeight;
+        }
+    };
+
+    /**
+     * Recalculates the layout
+     */
+    private updateLayout = (): void => {
+        if (isNil(this.rootElement.current) || isNil(this.itemContainerElement.current)) {
+            return;
+        }
+
+        if (this.props.orientation === Orientation.horizontal) {
+            this.updateDirection();
+        }
+
+        const scrollPos: number = this.getScrollPosition();
+        const lastIndex: number =
+            this.itemPositions.length === 0 ? 0 : this.itemPositions.length - 1;
+
+        const visibleRangeStartIndex: number = this.getThresholdItemIndex(
+            this.itemPositions,
+            0,
+            scrollPos
+        );
+        const visibleRangeEndIndex: number = this.getThresholdItemIndex(
+            this.itemPositions,
+            visibleRangeStartIndex,
+            scrollPos + this.viewportSpan
+        );
+
+        const bufferLength: number = this.getBufferLength();
+
+        let renderStartIndex: number = visibleRangeStartIndex - bufferLength;
+        if (renderStartIndex < 0) {
+            renderStartIndex = 0;
+        }
+
+        let renderEndIndex: number = visibleRangeEndIndex + bufferLength;
+        if (renderEndIndex > lastIndex) {
+            renderEndIndex = lastIndex;
+        }
+
+        window.clearTimeout(this.scrollLayoutUpdateTimer);
+
+        this.setState({
+            renderedRangeStartIndex: renderStartIndex,
```

Comment:
nit: suggest you eliminate the `renderStartIndex` variable and do something like:

```ts
(function(test: number, boundary: number) { return test > boundary ? test : boundary })(visibleRangeStartIndex - bufferLength, 0)
```

perhaps doing something similar for `renderEndIndex`.

Reply:
Better:
 this.setState({
            renderedRangeStartIndex: Math.max(visibleRangeStartIndex - bufferLength, 0),
            renderedRangeEndIndex: Math.min(visibleRangeEndIndex + bufferLength, lastIndex),
            isScrollable: this.itemContainerSpan > this.viewportSpan,
        });